File: src/OfferSB/src/test/java/com/example/offersb/test/OfferControllerTest.java
package com.example.offersb.test;

import com.example.offersb.Dtos.CreateSingleOfferRequestDto;
import com.example.offersb.controllers.OfferController;
import com.example.offersb.models.*;
import com.example.offersb.repositories.IOfferRepository;
import com.example.offersb.repositories.IProductRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.UUID;

@SpringBootTest
public class OfferControllerTest {

    @Autowired
    private IOfferRepository offerRepository;

    @Autowired
    private IProductRepository productRepository;

    @BeforeEach
    public void cleanDatabase() {
        offerRepository.deleteAll();
        productRepository.deleteAll();

        setupData();
    }

    private void setupData() {
        var products = new ArrayList<Product>();
        products.add(Product.create(UUID.fromString("fd239078-28ff-4158-a79c-ad533ce26dc5"), "Product 1", "Product 1 description"));
        products.add(Product.create(UUID.fromString("b7c196c6-e082-460f-af7f-0262f960d2ee"), "Product 2", "Product 2 description"));
        products.add(Product.create(UUID.fromString("a7948abe-571c-4e9d-a7f3-94a88ed10125"), "Product 3", "Product 3 description"));
        productRepository.saveAll(products);

        var offers = new ArrayList<Offer>();
        var price1 = Price.CreateFromGross(119, 19, "EUR");
        offers.add(SingleOffer.create(UUID.fromString("ad8b5807-cab4-4006-9425-1596bcf55c4a"), "Offer 1", price1, LocalDateTime.now(), LocalDateTime.now(), products.get(0)));

        var price2 = Price.CreateFromGross(99, 19, "EUR");
        offers.add(PackageOffer.create(UUID.fromString("d4495823-04a4-4ff9-9cc9-f17adee419f7"), "Offer 2", price2, LocalDateTime.now(), LocalDateTime.now(), products));
        offerRepository.saveAll(offers);
    }

    @Test
    public void testgetAllOffers_Should_Return_All_Offers() {
        // arrange
        var controller = new OfferController(offerRepository, productRepository);

        // act
        var result = controller.getAllOffers();

        // assert
        assert result != null;
        assert result.size() == 2;
    }

    @Test
    public void test_CreateSingleOffer_Should_be_Persisted() {
        // arrange
        var controller = new OfferController(offerRepository, productRepository);
        var offerDto = new CreateSingleOfferRequestDto("Offer 1", 119, 19, "EUR", LocalDateTime.now(), LocalDateTime.now(), UUID.fromString("fd239078-28ff-4158-a79c-ad533ce26dc5"));
        // act
        var result = controller.createSingleOffer(offerDto);

        // assert
        assert result != null;
        var persistedOffer = offerRepository.findById(result.getId());
        assert persistedOffer.isPresent();
        assert persistedOffer.get().getName().equals("Offer 1");
    }
}

File: src/Offering/OfferingTest/FixtureTests/OfferControllerTest.cs
using Microsoft.AspNetCore.Mvc;
using Offering.Controllers;
using Offering.Dtos;
using Offering.Models;
using Offering.Repositories;

namespace OfferingTest;

public class OfferControllerTest : IClassFixture<TestDatabaseFixture>, IDisposable
{
    public TestDatabaseFixture Fixture { get; }

    public OfferControllerTest(TestDatabaseFixture fixture)
    {
        Fixture = fixture;
    }

    public void Dispose()
    {
        Fixture.Cleanup();
    }

    [Fact]
    public async void ListOffers_Should_Return_All_Offers()
    {
        // Arrange
        Fixture.CreateMockDbData();
        var context = Fixture.CreateContext();
        var _offerRepository = new OfferRepository(context);
        var _productRepository = new ProductRepository(context);
        var _localisationRepository = new LocalizationRepository(context);
        var controller = new OfferController(_offerRepository, _productRepository, _localisationRepository);
        
        // Act
        var result = await controller.ListOffers();
        
        // Assert
        var okResult = result as OkObjectResult;
        var value = okResult.Value as List<Offer>;
        Assert.Equal(1, value.Count);
    }

    [Fact]
    public async void CreateSingleOffer_Should_Be_Persisted()
    {
        // Arrange
        var context = Fixture.CreateContext();
        var _offerRepository = new OfferRepository(context);
        var _productRepository = new ProductRepository(context);
        var _localisationRepository = new LocalizationRepository(context);
        context.Localizations.Add(Localization.Create("DE", "Germany", "Deutschland", "EUR"));
        context.SaveChanges();
        var controller = new OfferController(_offerRepository, _productRepository, _localisationRepository);
        var product = Product.Create(Guid.NewGuid(), "Product 2", "Description");
        await _productRepository.Add(product);
        var createSingleOfferRequestDto = new CreateSingleOfferRequestDto("Offer 2", 10, 19, DateTime.Now, DateTime.Now.AddDays(1), product.Id);

        // Act
        var result = await controller.CreateSingleOffer(createSingleOfferRequestDto);
        
        // Assert
        var okResult = result as OkObjectResult;
        var value = okResult.Value as SingleOffer;
        Assert.Equal("Offer 2", value.Name);
        var offer = context.Offers.Single(o => o.Name == "Offer 2");
        Assert.Equal(offer.Name, value.Name);
    }
    
}